File: ZemljaSlova.Services/EventService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MapsterMapper;
using Microsoft.EntityFrameworkCore;
using ZemljaSlova.Model;
using ZemljaSlova.Model.Requests;
using ZemljaSlova.Model.SearchObjects;
using ZemljaSlova.Services.Database;
using Microsoft.AspNetCore.Http;
using System.IO;

namespace ZemljaSlova.Services
{
    public class EventService : BaseCRUDService<Model.Event, EventSearchObject, Database.Event, EventUpsertRequest, EventUpsertRequest>, IEventService
    {
        private readonly IMapper _mapper;
        private readonly ITicketTypeService _ticketTypeService;

        public EventService(_200036Context context, IMapper mapper, ITicketTypeService ticketTypeService) : base(context, mapper)
        {
            _mapper = mapper;
            _ticketTypeService = ticketTypeService;
        }

        public async Task<Model.Event> GetEventWithTicketTypes(int id)
        {
            var entity = await Context.Events.Include(e => e.TicketTypes).FirstOrDefaultAsync(e => e.Id == id);
            
            if (entity == null)
                return null;
                
            var eventModel = _mapper.Map<Model.Event>(entity);
            
            // Calculate current quantities for each ticket type
            foreach (var ticketType in eventModel.TicketTypes)
            {
                ticketType.CurrentQuantity = await _ticketTypeService.GetCurrentQuantityAsync(ticketType.Id);
            }
            
            return eventModel;
        }

        public override IQueryable<Database.Event> AddFilter(EventSearchObject search, IQueryable<Database.Event> query)
        {
            if (search.IsTicketTypeIncluded == true)
            {
                query = query.Include("TicketTypes");
            }

            if (!string.IsNullOrEmpty(search.Name))
            {
                query = query.Where(e => e.Title.ToLower().Contains(search.Name.ToLower()));
            }

            // Filter out events that have already ended (show only current and future events)
            // Unless ShowPastEvents is explicitly set to true
            if (!search.ShowPastEvents.HasValue || !search.ShowPastEvents.Value)
            {
                var today = DateTime.Today;
                query = query.Where(e => e.StartAt >= today);
            }

            // Price range filters
            if (search.MinPrice.HasValue)
            {
                query = query.Where(e => e.TicketTypes.Any(t => t.Price >= search.MinPrice.Value));
            }

            if (search.MaxPrice.HasValue)
            {
                query = query.Where(e => e.TicketTypes.Any(t => t.Price <= search.MaxPrice.Value));
            }

            // Date range filters
            if (search.StartDateFrom.HasValue)
            {
                query = query.Where(e => e.StartAt >= search.StartDateFrom.Value);
            }

            if (search.StartDateTo.HasValue)
            {
                query = query.Where(e => e.StartAt <= search.StartDateTo.Value);
            }

            // Apply sorting
            if (!string.IsNullOrEmpty(search.SortBy))
            {
                switch (search.SortBy.ToLower())
                {
                    case "title":
                        query = search.SortOrder?.ToLower() == "desc" 
                            ? query.OrderByDescending(e => e.Title)
                            : query.OrderBy(e => e.Title);
                        break;
                    case "price":
                        // For price sorting, we need to join with ticket types and get min/max price
                        query = search.SortOrder?.ToLower() == "desc" 
                            ? query.OrderByDescending(e => e.TicketTypes.Max(t => t.Price))
                            : query.OrderBy(e => e.TicketTypes.Min(t => t.Price));
                        break;
                    case "date":
                        query = search.SortOrder?.ToLower() == "desc" 
                            ? query.OrderByDescending(e => e.StartAt)
                            : query.OrderBy(e => e.StartAt);
                        break;
                    default:
                        query = query.OrderByDescending(e => e.StartAt);
                        break;
                }
            }
            else
            {
                query = query.OrderByDescending(e => e.StartAt);
            }

            return base.AddFilter(search, query);
        }

        public override void BeforeDelete(Database.Event entity)
        {
            // Check if event has sold tickets
            var hasSoldTickets = Context.TicketTypeTransactions
                .Any(ttt => ttt.TicketType.EventId == entity.Id);
            
            if (hasSoldTickets)
            {
                throw new UserException("Nije moguće izbrisati događaj koji ima prodane karte.");
            }
        }

        public async Task<Model.Event> CreateEventFromForm(IFormCollection form)
        {
            // Extract form data
            string title = form["title"].FirstOrDefault() ?? "";
            string description = form["description"].FirstOrDefault() ?? "";
            string? location = form["location"].FirstOrDefault();
            string? organizer = form["organizer"].FirstOrDefault();
            string? lecturers = form["lecturers"].FirstOrDefault();
            string? maxPeopleStr = form["maxNumberOfPeople"].FirstOrDefault();
            int? maxNumberOfPeople = null;
            if (!string.IsNullOrEmpty(maxPeopleStr) && int.TryParse(maxPeopleStr, out int maxPeople))
            {
                maxNumberOfPeople = maxPeople;
            }

            // Parse dates
            DateTime startAt = DateTime.Now;
            if (DateTime.TryParse(form["startAt"].FirstOrDefault(), out DateTime parsedStartAt))
            {
                startAt = parsedStartAt;
            }

            DateTime endAt = DateTime.Now.AddHours(1);
            if (DateTime.TryParse(form["endAt"].FirstOrDefault(), out DateTime parsedEndAt))
            {
                endAt = parsedEndAt;
            }

            // Handle image file
            byte[] imageBytes = null;
            if (form.Files.Count > 0 && form.Files[0].Length > 0)
            {
                var imageFile = form.Files[0];
                using (var memoryStream = new MemoryStream())
                {
                    await imageFile.CopyToAsync(memoryStream);
                    imageBytes = memoryStream.ToArray();
                }
            }

            // Create event
            var eventEntity = new Database.Event
            {
                Title = title,
                Description = description,
                Location = location,
                StartAt = startAt,
                EndAt = endAt,
                Organizer = organizer,
                Lecturers = lecturers,
                CoverImage = imageBytes,
                MaxNumberOfPeople = maxNumberOfPeople
            };

            Context.Events.Add(eventEntity);
            await Context.SaveChangesAsync();

            return _mapper.Map<Model.Event>(eventEntity);
        }

        public async Task<Model.Event> UpdateEventFromForm(int id, IFormCollection form)
        {
            var eventEntity = await Context.Events.FirstOrDefaultAsync(e => e.Id == id);

            if (eventEntity == null)
            {
                throw new UserException("Događaj nije pronađen.");
            }

            // Extract form data
            string title = form["title"].FirstOrDefault() ?? "";
            string description = form["description"].FirstOrDefault() ?? "";
            string? location = form["location"].FirstOrDefault();
            string? organizer = form["organizer"].FirstOrDefault();
            string? lecturers = form["lecturers"].FirstOrDefault();
            string? maxPeopleStr = form["maxNumberOfPeople"].FirstOrDefault();
            int? maxNumberOfPeople = null;
            if (!string.IsNullOrEmpty(maxPeopleStr) && int.TryParse(maxPeopleStr, out int maxPeople))
            {
                maxNumberOfPeople = maxPeople;
            }

            // Parse dates
            if (DateTime.TryParse(form["startAt"].FirstOrDefault(), out DateTime parsedStartAt))
            {
                eventEntity.StartAt = parsedStartAt;
            }

            if (DateTime.TryParse(form["endAt"].FirstOrDefault(), out DateTime parsedEndAt))
            {
                eventEntity.EndAt = parsedEndAt;
            }

            // Update event data
            eventEntity.Title = title;
            eventEntity.Description = description;
            eventEntity.Location = location;
            eventEntity.Organizer = organizer;
            eventEntity.Lecturers = lecturers;
            eventEntity.MaxNumberOfPeople = maxNumberOfPeople;

            // Handle image file
            if (form.Files.Count > 0 && form.Files[0].Length > 0)
            {
                var imageFile = form.Files[0];
                using (var memoryStream = new MemoryStream())
                {
                    await imageFile.CopyToAsync(memoryStream);
                    eventEntity.CoverImage = memoryStream.ToArray();
                }
            }

            await Context.SaveChangesAsync();

            return _mapper.Map<Model.Event>(eventEntity);
        }
    }
}

File: UI/zemljaslova_mobile/lib/services/event_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import '../models/event.dart';
import '../models/ticket_type.dart';
import 'api_service.dart';

class EventService {
  final ApiService _apiService;
  
  EventService(this._apiService);
  
  Future<Map<String, dynamic>> fetchEvents({
    bool isTicketTypeIncluded = true,
    int? page,
    int? pageSize,
    String? name,
    String? sortBy,
    String? sortOrder,
    Map<String, String>? filters,
  }) async {
    try {
      List<String> queryParams = ['IsTicketTypeIncluded=$isTicketTypeIncluded'];
      
      if (page != null) {
        queryParams.add('Page=$page');
      }
      
      if (pageSize != null) {
        queryParams.add('PageSize=$pageSize');
      }
      
      if (name != null && name.isNotEmpty) {
        queryParams.add('Name=${Uri.encodeComponent(name)}');
      }
      
      if (sortBy != null && sortBy.isNotEmpty) {
        queryParams.add('SortBy=${Uri.encodeComponent(sortBy)}');
      }
      
      if (sortOrder != null && sortOrder.isNotEmpty) {
        queryParams.add('SortOrder=${Uri.encodeComponent(sortOrder)}');
      }
      
      // Add filters
      if (filters != null) {
        for (final entry in filters.entries) {
          queryParams.add('${entry.key}=${Uri.encodeComponent(entry.value)}');
        }
      }
      
      final queryString = queryParams.join('&');
      final response = await _apiService.get('Event?$queryString');
      
      debugPrint('API response: $response');
      
      if (response != null) {
        final eventsList = response['resultList'] as List;
        final totalCount = response['count'] as int;
        
        final events = eventsList
            .map((eventJson) => _mapEventFromBackend(eventJson))
            .toList();
            
        return {
          'events': events,
          'totalCount': totalCount,
        };
      }
      
      return {
        'events': <Event>[],
        'totalCount': 0,
      };
    } catch (e) {
      debugPrint('Failed to fetch events: $e');
      return {
        'events': <Event>[],
        'totalCount': 0,
      };
    }
  }
  
  Future<Event> getEventById(int id) async {
    try {
      final response = await _apiService.get('Event/GetEventWithTicketTypes/$id');
      
      if (response != null) {
        return _mapEventFromBackend(response);
      }
      
      throw Exception('Event not found');
    } catch (e) {
      debugPrint('Failed to get event: $e');
      throw Exception('Failed to get event: $e');
    }
  }

  Event _mapEventFromBackend(dynamic eventData) {
    String? coverImageUrl;
    if (eventData['coverImage'] != null) {
      if (eventData['coverImage'] is List) {
        final bytes = List<int>.from(eventData['coverImage']);
        coverImageUrl = 'data:image/jpeg;base64,${base64Encode(bytes)}';
      } else if (eventData['coverImage'] is String) {
        coverImageUrl = eventData['coverImage'];
      }
    }

    // Parse dates
    DateTime startAt = DateTime.now();
    if (eventData['startAt'] != null) {
      startAt = DateTime.parse(eventData['startAt']);
    }
    
    DateTime endAt = DateTime.now();
    if (eventData['endAt'] != null) {
      endAt = DateTime.parse(eventData['endAt']);
    }
    
    // Parse ticket types
    List<TicketType>? ticketTypes;
    
    if (eventData.containsKey('ticketTypes') && eventData['ticketTypes'] != null) {
      final typesList = eventData['ticketTypes'] as List;
      
      if (typesList.isNotEmpty) {
        ticketTypes = typesList.map((ticketTypeJson) {
          var id = ticketTypeJson['id'];
          var eventId = ticketTypeJson['eventId'];
          var price = ticketTypeJson['price'];
          var name = ticketTypeJson['name'];
          var description = ticketTypeJson['description'];
          var initialQuantity = ticketTypeJson['initialQuantity'];
          var currentQuantity = ticketTypeJson['currentQuantity'];
          
          return TicketType(
            id: id,
            eventId: eventId,
            price: (price as num).toDouble(),
            name: name,
            description: description,
            initialQuantity: initialQuantity,
            currentQuantity: currentQuantity,
          );
        }).toList();
      }
    }

    return Event(
      id: eventData['id'] ?? 0,
      title: eventData['title'] ?? '',
      description: eventData['description'] ?? '',
      location: eventData['location'],
      startAt: startAt,
      endAt: endAt,
      organizer: eventData['organizer'],
      lecturers: eventData['lecturers'],
      coverImageUrl: coverImageUrl,
      maxNumberOfPeople: eventData['maxNumberOfPeople'],
      ticketTypes: ticketTypes,
    );
  }
} 